### PR TITLE
Update to use golang 1.12

### PR DESCRIPTION
**Pull request by cblecker**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
/kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Update to use golang 1.12
```

### CODE REVIEW

**Comment by javier-b-perez** [comment]
Created kube-cross
k8s.gcr.io/kube-cross:v1.12.0-1

**Comment by cblecker** [comment]
/assign @liggitt 
for approval of apimachinery test fix

File: staging/src/k8s.io/apimachinery/pkg/util/naming/from_stack_test.go [hunk 1]
```diff
@@ -31,12 +31,12 @@ func TestGetNameFromCallsite(t *testing.T) {
...
 			name:            "ignore-package",
 			ignoredPackages: []string{"k8s.io/apimachinery/pkg/util/naming"},
-			expected:        "testing/testing.go:827",
+			expected:        "testing/testing.go:865",
 		},
 		{
 			name:            "ignore-file",
 			ignoredPackages: []string{"k8s.io/apimachinery/pkg/util/naming/from_stack_test.go"},
-			expected:        "testing/testing.go:827",
+			expected:        "testing/testing.go:865",
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/naming/from_stack_test.go]: this is a wacky test

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/naming/from_stack_test.go]: simplified in https://github.com/kubernetes/kubernetes/pull/74735 to not need rework on go version bumps

Comment (cblecker) [reply staging/src/k8s.io/apimachinery/pkg/util/naming/from_stack_test.go]: removed the change from here, as #74735 is way better.

File: build/root/WORKSPACE [hunk 2]
```diff
@@ -31,16 +31,16 @@ http_archive(
...
+    urls = mirror("https://github.com/bazelbuild/rules_go/releases/download/0.17.1/rules_go-0.17.1.tar.gz"),
 )
 
 load("@io_bazel_rules_go//go:deps.bzl", "go_register_toolchains", "go_rules_dependencies")
 
 go_rules_dependencies()
 
 go_register_toolchains(
-    go_version = "1.11.5",
+    go_version = "1.12",
```
Comment (javier-b-perez) [inline build/root/WORKSPACE]: should this be 1.12.0?

**Comment by spiffxp** [comment]
FWIW I'm inclined to wait on this until v1.15 unless someone can give a strong case for why this is needed so close to code freeze.  I'm open to suggestion, but feedback I got from @kubernetes/sig-scalability-pr-reviews was this wasn't pressing to them from a performance perspective.

/cc @wojtek-t

**Comment by liggitt** [comment]
> FWIW I'm inclined to wait on this until v1.15 unless someone can give a strong case for why this is needed so close to code freeze

the http/2 issue (c.f. https://github.com/kubernetes/kubernetes/issues/74412, https://github.com/golang/go/issues/27044) would be resolved by this, and is significant

**Comment by ixdy** [comment]
another thought, re: https://github.com/kubernetes/kubernetes/issues/73286: go 1.11 will fall out of its support window (~end of August) before k8s 1.14 falls out of its support window (~October).

**Comment by spiffxp** [comment]
/cc @shyamjvs 
OK, I'm feeling convinced to at least give this a shot if the tests are all passing.  Pinged in #sig-scalability

**Comment by liggitt** [comment]
both of those seem like significant reasons to me.

File: build/root/WORKSPACE (hunk 2 above)
Comment (cblecker) [reply build/root/WORKSPACE]: Nope. It needs to match an available toolchain: https://github.com/bazelbuild/rules_go/blob/2728c534ae1bce3c073157370006168ae3b09f2c/go/toolchain/toolchains.bzl#L31

**Comment by liggitt** [comment]
from https://github.com/kubernetes/kubernetes/issues/74412#issuecomment-468523544:

> it looks like since we vendor golang.org/x/net/http2, we actually have to bump our vendored copy to the go1.12 level to pick up the fix. I was able to write a simple integration test against the cached secret manager. On master, it consistently failed after establishing 236 watches. With our vendored http2 copy bumped to go1.12 levels, it worked all the way up to 10,000 distinct watches (which is where I stopped checking)

**Comment by cblecker** [comment]
@liggitt yup, I can easily do that once this merges.

**Comment by BenTheElder** [comment]
NOTE: those appear to be normal flakes, not related to this PR, but the govet failure is legitimate:
```
# k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm
cmd/kubeadm/app/apis/kubeadm/types.go:41:2: struct field Kind repeats json tag "kind" also at types.go:48
cmd/kubeadm/app/apis/kubeadm/types.go:41:2: struct field APIVersion repeats json tag "apiVersion" also at types.go:55
# k8s.io/kubernetes/pkg/controller/testutil
pkg/controller/testutil/test_utils.go:377:2: missing ... in args forwarded to printf-like function
# k8s.io/kubernetes/pkg/kubelet/container
pkg/kubelet/container/sync_result_test.go:48:3: Errorf format %q has arg result of wrong type k8s.io/kubernetes/pkg/kubelet/container.PodSyncResult
pkg/kubelet/container/sync_result_test.go:56:3: Errorf format %q has arg result of wrong type k8s.io/kubernetes/pkg/kubelet/container.PodSyncResult
pkg/kubelet/container/sync_result_test.go:66:3: Errorf format %q has arg result of wrong type k8s.io/kubernetes/pkg/kubelet/container.PodSyncResult
# k8s.io/kubernetes/pkg/kubelet/kuberuntime
pkg/kubelet/kuberuntime/kuberuntime_manager_test.go:387:3: Errorf format %q has arg expected of wrong type []*k8s.io/kubernetes/pkg/kubelet/container.Pod
# k8s.io/kubernetes/pkg/registry/core/service/allocator
pkg/registry/core/service/allocator/utils_test.go:49:4: Errorf format %d has arg test.n of wrong type *math/big.Int
make[1]: *** [vet] Error 1
```

**Comment by cblecker** [comment]
Ugh. Okay maybe not on the updating golang.org/x/net:
https://github.com/kubernetes/kubernetes/pull/71501#issuecomment-468547962

Godeps is the worst.

**Comment by neolit123** [comment]
> k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm
cmd/kubeadm/app/apis/kubeadm/types.go:41:2: struct field Kind repeats json tag "kind" also at types.go:48
cmd/kubeadm/app/apis/kubeadm/types.go:41:2: struct field APIVersion repeats json tag "apiVersion" also at types.go:55

just saw that in another PR after test-infra updated.
this is "a bit" of a blocker.
cc @fabriziopandini @timothysc @rosti

and of course govet is right about the kubeadm embedding of a GVKed config in another GVKed "situation". i've mentioned this is a bad idea, at least once.

the rest of the errors seem easier to manage.

**Comment by BenTheElder** [comment]
[rolled back test-infra changes due to the `go vet` failures https://github.com/kubernetes/test-infra/pull/11575]

**Comment by spiffxp** [comment]
Based on the PR trail I'm a bit confused, are we using golang 1.12 here but not in test-infra?

On the go vet failures, how could we verify we've addressed them via presubmits/PRs instead of merging things?

**Comment by neolit123** [comment]
@spiffxp 
the k/k govet failure fixes are WIP.
we need to update test infra again to 1.12 and test them and to make sure the verify job passes on k/k.

**Comment by spiffxp** [comment]
Status update:
- there have been green runs of https://testgrid.k8s.io/sig-release-master-informing#gce-master-scale-performance since this merged
- https://testgrid.k8s.io/sig-release-master-informing#gce-master-scale-correctness has yet to have a green run since we've merged
- https://github.com/kubernetes/test-infra/issues/11594 is open to see why prow is triggering more scalability jobs than intended, causing us to run out of quota

So at this point I'm not seeing a "no, revert, go1.12 broke things" but neither am I seeing a "yes, things look green since we've moved to go1.12".  At the moment I'm inclined to wait and see.

WDYT @wojtek-t @shyamjvs ?

**Comment by shyamjvs** [comment]
Yeah, seems like there were green runs after this pr merged and they look fine at a high-level. There are recent red runs (quite likely unrelated). Also, it's better to wait for a green correctness run before deciding.

**Comment by mm4tt** [comment]
Both https://testgrid.k8s.io/sig-release-master-informing#gce-master-scale-performance and https://testgrid.k8s.io/sig-release-master-informing#gce-master-scale-correctness are directly affected by kubernetes/test-infra#11594. 
As long as the issue is not fixed the probability of the tests passing is rather low.

**Comment by mm4tt** [comment]
The gce-master-scale-correctness failure is tracked in https://github.com/kubernetes/kubernetes/issues/75095

**Comment by spiffxp** [comment]
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-correctness/219 - most recent run was green

I feel like this gives me sufficient confidence we're going to proceed with golang 1.12.  Any objections @shyamjvs @wojtek-t?

**Comment by wojtek-t** [comment]
gce-scale-correctness isn't enough signal in my opinion - we should wait for gce-scale-performance

**Pull request merged**